docs(replica): add replica rebuilding flow (#1161)
longhorn/website: content/docs/1.10.0/advanced-resources/rebuilding/_index.md

### Full Replica Rebuilding

If the replica is unrecoverable or replenished, Longhorn will synchronize the whole data from the healthy replica. Reconstruct the replica by sending the full snapshot chain.

Full replica rebuilding will waste network bandwidth and generate numerous disk WRITE operations on the target replica node, but it is the only way to rebuild a replica if there is no data on the target node.

### Delta Replica Rebuilding

- This is available for failed replica reusage only, and there is an existing snapshot file (with the same name) in the failed replica data directory
- There is no checksum/checksum file for the snapshot file. It's like a fallback of fast rebuilding

- **Pros**:
  - Reduce network bandwidth consumption.
- **Cons**:
  - Increased CPU overhead.
  - Rebuilding time is influenced by CPU performance.

### Fast Replica Rebuilding

Fast rebuilding is enabled when:

- The fast replica building setting is enabled:
  - `fast-replica-rebuild-enabled: true`
- Snapshot checksum files are created:
  - `snapshot-data-integrity`: `enabled` or

>   - `snapshot-data-integrity-immediate-check-after-snapshot-creation: true`

Reviewer: Elaborate more on
- when a checksum of a snapshot is calculated.

Author: Added some descriptions and added a section to analyze the setting trade-off.

## Factors That Affect Rebuilding Performance

- Large volume head or No snapshot exists
  - **WHY**:  
    - If the replica fails and rebuilding starts immediately, there is no opportunity to generate a checksum file for this large volume head. Then Longhorn has to do a full rebuild for all data in this large volume head
    - If the replica soonly fails and gets rebuilt after the snapshot of this large volume head is created, there is not enough time to calculate the checksum. As a result, Longhorn has to do a delta rebuilding for all data in this newly created snapshot, which will cause checksum calculation overhead and numerous checksum comparisons between the target replica and the source replica
  - **HOW**:
    1. Enable `snapshot-data-integrity-immediate-check-after-snapshot-creation` or `snapshot-data-integrity` setting, so checksums are precomputed.
    2. Take regularly snapshots by a recurring job.
- Snapshot Purged
  - **WHY**:  
  When the snapshot purged of volume starts, the system-generated snapshot will be coalesced to the next snapshot, and the checksum of the next snapshot will be invalid.
  - **HOW**:

>     1. Enable `snapshot-data-integrity-immediate-check-after-snapshot-creation` to ensure the checksum precomputed after the snapshot purged.

Reviewer: Please elaborate the trade-off the setting `snapshot-data-integrity-immediate-check-after-snapshot-creation`.

Author: Added a section to analyze the setting trade-off. Thanks.

Reviewer: Please briefly mention the trade-off here as well. it will be more clear for users.

Author: Added. Thanks.

    2. Create a snapshot proactively and wait for the checksum to be computed before upgrades or rebuild events
- Multiple replica failures
  - **WHY**:  
  Increases rebuild complexity and duration. With the `auto-cleanup-system-generated-snapshot` setting `true` and no user-created snapshots, if two replicas fail before either has the chance to be rebuilt, at least one `full data transfer` is required to bring the volume back to a healthy state.  
  For more details, [Avoid "full data transfer" when rebuilding two failed replicas
](https://github.com/longhorn/longhorn/issues/9335)
  - **HOW**:

>     1. Manually disable auto-cleanup-system-generated-snapshot before doing maintenance.

Reviewer: ```suggestion
    1. Manually disable `auto-cleanup-system-generated-snapshot before doing maintenance`.
```

---
title: Replica Rebuilding
weight: 6
---
- [Full Replica Rebuilding](#full-replica-rebuilding)
- [Delta Replica Rebuilding](#delta-replica-rebuilding)
- [Fast Replica Rebuilding](#fast-replica-rebuilding)
- [Factors That Affect Rebuilding Performance](#factors-that-affect-rebuilding-performance)
- [Relevant Settings](#relevant-settings)

When Longhorn detects a failed or deleted replica, it automatically triggers a rebuilding process to rebuild the failed replica. This document outlines the replica rebuilding flow, covering full rebuilding, delta rebuilding, and fast rebuilding. It also explains the limitations of each rebuilding method.

Replica rebuilding may be triggered under the following scenarios:

- Node reboot or eviction.
- A replica becomes unhealthy or deleted.

>  {{< figure alt="Replica Rebuilding Flow Diagram" src="/img/diagrams/architecture/replica-rebuilding-flow.png" >}}

Author: Modified. Thanks

            - If **YES**, Longhorn skips the corresponding snapshot data transfer, which reduces the CPU usage, disk IO, network IO, and rebuilding time. [Fast Replica Rebuilding](#fast-replica-rebuilding)
            - If **NO**, Longhorn calculates and compares block-by-block checksums for the snapshot file between the target replica and the source replica. If there is a mismatch, the corresponding data block will be synced. [Delta Replica Rebuilding](#delta-replica-rebuilding)

## Full Replica Rebuilding

If the replica is unrecoverable or replenished, Longhorn will synchronize the whole data from the healthy replica. Reconstruct the replica by sending the full snapshot chain.

Full replica rebuilding will waste network bandwidth and generate numerous disk WRITE operations on the target replica node, but it is the only way to rebuild a replica if there is no data on the target node.

## Delta Replica Rebuilding

6. Compare block-by-block checksums if [Fast Replica Rebuilding](#fast-replica-rebuilding) is not enabled. Longhorn engine will compare each snapshot files block-by-block between the target replica and the source replica. The checksum is computed using SHA-512 method.

> 7. If local and remote checksums do not match, receiver server of the target replica will synchronize missing or mismatched data.

Author: Added and modified. Thanks

- **Pros**:
  - Reduce network bandwidth consumption.

>   - Reduce disk IO.

Reviewer 2: I am not sure if we need to mention this in **Cons**. WDYT? @derekbit

>   - Reduce disk IO.

Reviewer: Agree with @shuo-wu

- **Cons**:
  - Increased CPU overhead.
  - Rebuilding time is influenced by CPU performance.

## Fast Replica Rebuilding

Fast rebuilding is enabled when:

- The fast replica building setting is enabled:
  - `fast-replica-rebuild-enabled: true`
- Snapshot checksum files are created:
  - `snapshot-data-integrity`: `enabled` or
  - `snapshot-data-integrity-immediate-check-after-snapshot-creation: true`

6. Skip checking unchanged snapshots if the checksums are the same in the checksum files of snapshots.
7. If the value of snapshot checksum files between the target replica and the source replica does not exist or is not identical, it will go to the delta rebuilding flow.

- **Pros**:
  - Minimize network bandwidth consumption.
  - Minimize disk IO.
- **Cons**:
  - Calculating the snapshot checksum can be time-consuming.

The snapshot checksum is computed:

- Immediately after the snapshot is created with the `snapshot-data-integrity-immediate-check-after-snapshot-creation` setting enabled.
- Regularly (by default: 7 days) if the `snapshot-data-integrity` setting is `enabled`.

For more details, [Fast Replica Rebuilding](./fast-replica-rebuilding).

> ## Factors That Affect Rebuilding Performance

Reviewer 2: Besides, I am not sure if we can mention the concurrent rebuilding setting in this section

Author: Removed. Thanks

- Large volume head or No snapshot exists
  - **WHY**:  

>   It will take a snapshot without the checksum for the large volume-head image, and it will cause checksum calculation overhead and numerous checksum comparisons between the target replica and the source replica.

Reviewer 2: We can clarify more about this:
- If the replica failure and rebuilding happen immediately, there is no way to prepare a checksum file for this large volume head. Then Longhorn has to do a full rebuild for all data in this large volume head
- If the replica soonly fails and gets rebuilt after making this large volume head a snapshot, there is not enough time to calculate the checksum. Then Longhorn has to do a delta rebuild for all data in this newly created snapshot, which will cause checksum calculation overhead and numerous checksum comparisons between the target replica and the source replica

Author: Added and modified. Thanks

  - **HOW**:
    1. Enable `snapshot-data-integrity-immediate-check-after-snapshot-creation` or `snapshot-data-integrity` setting, so checksums are precomputed.
    2. Take regularly snapshots by a recurring job.
- Snapshot Purged
  - **WHY**:  
  When the snapshot purged of volume starts, the system-generated snapshot will be coalesced to the next snapshot, and the checksum of the next snapshot will be invalid.
  - **HOW**:
    1. Enable `snapshot-data-integrity-immediate-check-after-snapshot-creation` to ensure the checksum precomputed after the snapshot purged.
    2. Create a snapshot proactively and wait for the checksum to be computed before upgrades or rebuild events
- Concurrent rebuilds
  - **WHY**:  
  The number of concurrent rebuild on a node will occupy the CPU usage, disk IO, and network IO resources.
  - **HOW**:  
  Tune the number of concurrent rebuilding by the `concurrent-replica-rebuild-per-node-limit` setting on a node.

> - Multiple replica failures

Author: Keep it.

---
title: Replica Rebuilding
weight: 6
---
- [Replica Rebuilding Workflow](#replica-rebuilding-workflow)
  - [Full Replica Rebuilding](#full-replica-rebuilding)
  - [Delta Replica Rebuilding](#delta-replica-rebuilding)
  - [Fast Replica Rebuilding](#fast-replica-rebuilding)
- [Factors That Affect Rebuilding Performance](#factors-that-affect-rebuilding-performance)
- [Use Cases](#use-cases)
  - [Node Reboot During Upgrade](#node-reboot-during-upgrade)
  - [Short-Term Node Drain](#short-term-node-drain)
- [Relevant Settings](#relevant-settings)
  - [Settings Trade-Off Analysis](#settings-trade-off-analysis)

> When Longhorn detects a failed or deleted replica, it automatically triggers a rebuilding process to rebuild the failed replica. This document outlines the replica rebuilding flow, covering full rebuilding, delta rebuilding, and fast rebuilding. It also explains the limitations of each rebuilding method.
> 
> Rebuilding will not start when:
> 
> - The volume is migrating to another node.
> - The volume is an old restore/DR volume.
> - The volume is expanding in size.
> 
> ## Replica Rebuilding Workflow
> 
> Replica rebuilding may be triggered under the following scenarios:
> 
> - Node reboot or eviction.
> - A replica becomes unhealthy or deleted.

Reviewer 3: ```suggestion
When Longhorn detects a failed or deleted replica, it automatically initiates a rebuilding process. This document outlines the replica rebuilding workflow, including **full**, **delta**, and **fast** rebuilding methods. It also explains the limitations associated with each method.

**Rebuilding will not start in the following scenarios**:

- The volume is migrating to another node.
- The volume is an old restore/DR volume.
- The volume is expanding in size.

## Replica Rebuilding Workflow

Replica rebuilding may be triggered in the following scenarios:

- A node is rebooted or evicted.
- A replica becomes unhealthy or is deleted.
```

Author: Modified. Thanks.

5. For each snapshot synchronization,
    - Check whether the snapshot file exists in the target replica data directory
        - If **NO**, directly transfer all data from the source replica to the destination replica. [Full Replica Rebuilding](#full-replica-rebuilding)
        - If **YES**, check whether the value of snapshot checksum files between the target replica and the source replica exists and is identical:
            - If **YES**, Longhorn skips the corresponding snapshot data transfer, which reduces the CPU usage, disk IO, network IO, and rebuilding time. [Fast Replica Rebuilding](#fast-replica-rebuilding)
            - If **NO**, Longhorn calculates and compares block-by-block checksums for the snapshot file between the target replica and the source replica (The checksum is computed using SHA-512 method). If there is a mismatch, the corresponding data block will be synced. [Delta Replica Rebuilding](#delta-replica-rebuilding)

### Full Replica Rebuilding

If the replica is unrecoverable or needs to be replenished, Longhorn will synchronize the whole data from the healthy replica. It reconstructs the replica by sending the full snapshot chain.

> Full replica rebuilding will waste network bandwidth and generate numerous disk WRITE operations on the target replica node, but it is the only way to rebuild a replica if there is no data on the target node.

Author: Modified. Thanks.

  - **How to prevent**:
    1. Enable the `snapshot-data-integrity-immediate-check-after-snapshot-creation` or `snapshot-data-integrity` setting, so checksums are precomputed. Trade-off: it will consume CPU usage, disk IO resources when computing the checksums of snapshots.
    2. Take snapshots regularly using a recurring job.
- Snapshot Purged
  - **Why it matters**:  
  When snapshot purging of a volume starts, the system-generated snapshot will be coalesced to the next snapshot, and the checksum of the next snapshot will become invalid.
  - **How to prevent**:
    1. Enable the `snapshot-data-integrity-immediate-check-after-snapshot-creation` setting to ensure the checksum is precomputed after the snapshot is purged.
    2. Create a snapshot proactively and wait for the checksum to be computed before upgrades or rebuild events.
- Concurrent rebuilds
  - **Why it matters**:  
  The number of concurrent rebuilds on a node will consume CPU usage, disk IO, and network IO resources.
  - **How to prevent**:  
  Tune the number of concurrent rebuilds using the `concurrent-replica-rebuild-per-node-limit` setting on a node.

> - Multiple replica failures

Author: Modified. Thanks.

        - If **YES**, check whether the value of snapshot checksum files between the target replica and the source replica exists and is identical:
            - If **YES**, Longhorn skips the corresponding snapshot data transfer, which reduces the CPU usage, disk IO, network IO, and rebuilding time. [Fast Replica Rebuilding](#fast-replica-rebuilding)
            - If **NO**, Longhorn calculates and compares block-by-block checksums for the snapshot file between the target replica and the source replica (The checksum is computed using SHA-512 method). If there is a mismatch, the corresponding data block will be synced. [Delta Replica Rebuilding](#delta-replica-rebuilding)

### Full Replica Rebuilding

If the replica is unrecoverable or replenished, Longhorn will synchronize the whole data from the healthy replica. Reconstruct the replica by sending the full snapshot chain.

Full replica rebuilding will waste network bandwidth and generate numerous disk WRITE operations on the target replica node, but it is the only way to rebuild a replica if there is no data on the target node.

> ### Delta Replica Rebuilding

Reviewer: Please use 1-2 sentences to briefly explain how delta replica rebuilding works.

Author: Added. Thanks.

- This is available for failed replica reusage only, and there is an existing snapshot file (with the same name) in the failed replica data directory

> - There is no checksum/checksum file for the snapshot file. It's like a fallback of fast rebuilding

Reviewer: What does `no checksum/checksum file` mean?

- **Pros**:
  - Reduce network bandwidth consumption.
- **Cons**:

>   - Increased CPU overhead.

Reviewer: Why `Increased CPU overhead.`

  - Rebuilding time is influenced by CPU performance.

### Fast Replica Rebuilding

Fast rebuilding is enabled when:

- The fast replica building setting is enabled:
  - `fast-replica-rebuild-enabled: true`
- Snapshot checksum files are created:

>   - `snapshot-data-integrity`: `enabled`, and a scheduled job will calculate checksums of all snapshots in a configured interval or

Reviewer: ```suggestion
  - `snapshot-data-integrity` is set to `enabled`: a scheduled job will calculate checksums of all snapshots in a configured interval or
```

> a configured interval 

Mention which setting determines the interval

Author: Modified. Thanks.

>   - `snapshot-data-integrity-immediate-check-after-snapshot-creation: true`, calculate the snapshot checksum after the snapshot is created.

Reviewer: ```suggestion
  - `snapshot-data-integrity-immediate-check-after-snapshot-creation` is set to `true`, immediately calculate the snapshot checksum after the snapshot is created.
```

>   However, the checksum calculation negatively impacts the storage performance.

Reviewer: Please explain why the checksum calculation negatively impacts the storage performance.

Author: Added the descriptions. Thanks.

> 6. Skip checking unchanged snapshots if the checksums are the same in the checksum files of snapshots.

Reviewer: Not only check checksum but also check `ctime` of the file. Please walk through the source code and fix the step.

7. If the value of snapshot checksum files between the target replica and the source replica does not exist or is not identical, it will go to the delta rebuilding flow.

- **Pros**:
  - Minimize network bandwidth consumption.
  - Minimize disk IO.
- **Cons**:
  - Calculating the snapshot checksum can be time-consuming.
  - The timing of checksum calculation is unpredictable. It can be triggered even if the node is experiencing heavy IO load.

The snapshot checksum is computed:

- Immediately after the snapshot is created with the `snapshot-data-integrity-immediate-check-after-snapshot-creation` setting enabled.
- Regularly (by default: 7 days) if the `snapshot-data-integrity` setting is `enabled`.

> For more details, [Fast Replica Rebuilding](./fast-replica-rebuilding).

Reviewer: Seems redundant. Please consolidate it into the start of the section
```
- The fast replica building setting is enabled:
  - `fast-replica-rebuild-enabled: true`
- Snapshot checksum files are created:
  - `snapshot-data-integrity`: `enabled`, and a scheduled job will calculate checksums of all snapshots in a configured interval or
  - ....
```

Author: Merged. Thanks.

## Factors That Affect Rebuilding Performance

- Large volume head or No snapshot exists

>   - **WHY**:  

Reviewer: ```suggestion
  - **Why it matters**:  
```

Reviewer: Apply to others

    - If the replica fails and rebuilding starts immediately, there is no opportunity to generate a checksum file for this large volume head. Then Longhorn has to do a full rebuild for all data in this large volume head
    - If the replica quickly fails and gets rebuilt after the snapshot of this large volume head is created, there is not enough time to calculate the checksum. As a result, Longhorn has to do a delta rebuilding for all data in this newly created snapshot, which will cause checksum calculation overhead and numerous checksum comparisons between the target replica and the source replica

>   - **HOW**:

Reviewer: ```suggestion
  - **How to prevens**:
```

Author: Modified. Thanks.

||||

### Settings Trade-Off Analysis

- `fast-replica-rebuild-enabled`
  - `enabled`  
    Skip snapshot data integrity checks if the checksum is precomputed and up to date. Fast, but skipped snapshot data is not compared.
  - `disabled`  
    Always do delta replica rebuilding when a reusable failed replica is available. Slow, but the snapshot data matches exactly.
- [snapshot-data-integrity](../data-integrity/snapshot-data-integrity-check)
  - `enabled`  
  By default, it will compute the checksum of all snapshots every 7 days. Calculating checksums for all snapshots consumes worker node resources and adds processing time.
- `snapshot-data-integrity-cronjob`
  - `0 0 */7 * *`  
  If the `snapshot-data-integrity` setting is `enabled`, a cron job will be scheduled to start calculate all snapshot checksums. The snapshot checksum will not be computed during the configured interval. For example, by default, snapshots created within the last 7 days will not have their checksums precomputed.
- `snapshot-data-integrity-immediate-check-after-snapshot-creation`
  - `true`  

>   Calculate the snapshot checksum after the snapshot is created immediately. It will also consume worker node resources and adds unexpectedly processing time. If numerous snapshots are created every day, it will have a negative impact.

Reviewer: > If numerous snapshots are created every day, it will have a negative impact.

The sentence seems redundant.

Author: Removed. Thanks.

  - `false`  

>   By default, snapshots created within the last 7 days will not have their checksums precomputed if the `snapshot-data-integrity` is `enabled`. Therefore, if numerous snapshots are created every day, it will compare numerous data block of snapshot with calculating checksums during the failed replica rebuilding.

Reviewer: ```suggestion
  By default, snapshots created within the last 7 days determined by the setting `xxxxxxx` will not have their checksums precomputed if the `snapshot-data-integrity` is `enabled`. Therefore, if numerous snapshots are created every day, it will compare numerous data block of snapshot with calculating checksums during the failed replica rebuilding.
```

> Therefore, if numerous snapshots are created every day, it will compare numerous data block of snapshot with calculating checksums during the failed replica rebuilding.

Weird statement. This behavior is unrelated to the frequency of snapshot creation, such as “numerous snapshots are created every day…”.

You can say, like, `If the checksum of a snapshot file has not been calculated, Longhorn will fall back to delta rebuilding`.

Author: Modified. Thanks.

  - [Short-Term Node Drain](#short-term-node-drain)
- [Relevant Settings](#relevant-settings)
  - [Settings Trade-Off Analysis](#settings-trade-off-analysis)

When Longhorn detects a failed or deleted replica, it automatically initiates a rebuilding process. This document outlines the replica rebuilding workflow, including **full**, **delta**, and **fast** rebuilding methods. It also explains the limitations associated with each method.

**Rebuilding will not start in the following scenarios**:

- The volume is migrating to another node.
- The volume is an old restore/DR volume.
- The volume is expanding in size.

## Replica Rebuilding Workflow

Replica rebuilding may be triggered in the following scenarios:

- A node is rebooted or evicted.
- A replica becomes unhealthy or is deleted.

 {{< figure alt="Replica Rebuilding Flow Diagram" src="/img/diagrams/architecture/replica-rebuilding-flow.png" >}}

1. Mark the target replica with `WO` (Write-Only) mode.
2. Create a new snapshot to serve as the volume head reference point for data integrity checks.
3. Generate the synchronization file list for the volume head and snapshot files.

> 4. Launch a receiver server on the target replica for each snapshot, then instruct the source replica to begin data synchronization.

Reviewer 2: Do we need to mention the workflow differences for the v2 data engine?

Author: Added for `v1 data engine` in the summary description. Thanks

Reviewer 2: The rebuilding flows between v1 and v2 are similar in general. But there are still some minor differences. Can we explain the differences in this general rebuilding flow intro part?

Author: Split v1 and v2 data engines into two parts. Thanks.

5. For each snapshot synchronization:
    - Check whether the snapshot file exists in the target replica's data directory:
        - If **NO**, transfer the entire snapshot data from the source replica to the target replica. See [Full Replica Rebuilding](#full-replica-rebuilding)
        - If **YES**, check whether the snapshot checksum files exist and are identical between the target and source replicas:
            - If **YES**, Longhorn skips transferring that snapshot’s data. This optimization reduces CPU usage, disk I/O, network I/O, and overall rebuild time. See [Fast Replica Rebuilding](#fast-replica-rebuilding)

>             - If **NO**, Longhorn calculates and compares block-by-block checksums for the snapshot file using the SHA-512 algorithm. If mismatches are found, only the differing data blocks are synchronized. See [Delta Replica Rebuilding](#delta-replica-rebuilding)

Reviewer 2: Same. Do we need to mention the workflow differences for the v2 data engine?

Author: Added for `v1 data engine` in item 5 title. Thanks

### Full Replica Rebuilding

If the replica is unrecoverable or has no existing data, Longhorn synchronizes all data from a healthy replica. It reconstructs the replica by transferring the full snapshot chain.

Full replica rebuilding consumes significant network bandwidth and results in heavy disk write operations on the target node. However, it is required when the target replica has no usable data.

> ### Delta Replica Rebuilding

Reviewer 2: Do we need to emphasize that this is for v1 data engine only?

Author: Added `for v1 data engine` in the summary description. Thanks

  2. Based on the rebuild scenario:
      - **If fast replica rebuilding is enabled and all snapshot checksums exist:**
        [Fast Replica Rebuilding](#fast-replica-rebuilding) is triggered.
        - Only changed blocks in the volume head are synced, avoiding both full and delta rebuilding.
      - **If fast replica rebuilding is enabled but some snapshot checksums are missing:**
        [Delta Replica Rebuilding](#delta-replica-rebuilding) is used.
        - Changed blocks from snapshots without checksums are synced, avoiding full rebuilding.
      - **If fast replica rebuilding is disabled:**
        - Changed blocks of **all snapshots** are synced (delta rebuilding), avoiding full rebuilding.

### Short-Term Node Drain

If a worker node is drained for short-term maintenance and then quickly restored:

1. The replica on the drained node is marked as failed immediately.
2. If the node is uncordoned before `replica-replenishment-wait-interval` expires, Longhorn attempts to reuse the failed replica..
3. Rebuild behavior follows the same logic as described in the previous use case.

## Relevant Settings

| Setting| Default | Description |
| :--- | :---: | :--- |

> | `fast-replica-rebuild-enabled` | true | Enables fast replica rebuilding for the v1 data engine. Relies on precomputed snapshot checksums. |

Author: Modified. Thanks

> | `v2-data-engine-fast-replica-rebuilding` | false | Enables fast replica rebuilding for the v2 data engine. Also depends on snapshot checksums. |

Author: Removed. Thanks.

| [snapshot-data-integrity](../data-integrity/snapshot-data-integrity-check) | fast-check | Hashes snapshot disk files only if they are unhashed or their modification time has changed. |

> | `v2-data-engine-snapshot-data-integrity` | fast-check | Enables/disables snapshot hashing and data integrity checking for the v2 engine. |

Author: Removed. Thanks.

  2. Based on the rebuild scenario:
      - **If fast replica rebuilding is enabled and all snapshot checksums exist:**
        [Fast Replica Rebuilding](#fast-replica-rebuilding) is triggered.
        - Only changed blocks in the volume head are synced, avoiding both full and delta rebuilding.
      - **If fast replica rebuilding is enabled but some snapshot checksums are missing:**
        [Delta Replica Rebuilding](#delta-replica-rebuilding) is used.
        - Changed blocks from snapshots without checksums are synced, avoiding full rebuilding.
      - **If fast replica rebuilding is disabled:**
        - Changed blocks of **all snapshots** are synced (delta rebuilding), avoiding full rebuilding.

### Short-Term Node Drain

If a worker node is drained for short-term maintenance and then quickly restored:

1. The replica on the drained node is marked as failed immediately.
2. If the node is uncordoned before `replica-replenishment-wait-interval` expires, Longhorn attempts to reuse the failed replica..
3. Rebuild behavior follows the same logic as described in the previous use case.

## Relevant Settings

> | Setting| Default | Description |
> | :--- | :---: | :--- |
> | `fast-replica-rebuild-enabled` | true | Enables fast replica rebuilding. Relies on precomputed snapshot checksums. |
> | [snapshot-data-integrity](../data-integrity/snapshot-data-integrity-check) | fast-check | Hashes snapshot disk files only if they are unhashed or their modification time has changed. |
> | `snapshot-data-integrity-cronjob` | 0 0 */7 * * | Cron schedule to compute checksums for all snapshots (default: every 7 days). |
> | `snapshot-data-integrity-immediate-check-after-snapshot-creation` | false | If enabled, checksums are computed immediately after snapshot creation. |
> | `replica-replenishment-wait-interval` | 600 |  Time in seconds to wait before creating a new replica, allowing reuse of failed replicas. |
> | `concurrent-replica-rebuild-per-node-limit` | 5 | Limits the number of concurrent replica rebuilds per node. |
> | `offline-replica-rebuilding` | false | Determines if degraded replicas are rebuilt while the volume is detached. |
> ||||

Reviewer 3: ```suggestion
| Setting| Default | Description |
| :--- | :---: | :--- |
| `fast-replica-rebuild-enabled` | `true` | Enables fast replica rebuilding. Relies on precomputed snapshot checksums. |
| [`snapshot-data-integrity`](../data-integrity/snapshot-data-integrity-check) | `fast-check` | Hashes snapshot disk files only if they are unhashed or their modification time has changed. |
| `snapshot-data-integrity-cronjob` | `0 0 */7 * *` | Cron schedule to compute checksums for all snapshots (default: every 7 days). |
| `snapshot-data-integrity-immediate-check-after-snapshot-creation` | `false` | If enabled, checksums are computed immediately after snapshot creation. |
| `replica-replenishment-wait-interval` | `600` |  Time in seconds to wait before creating a new replica, allowing reuse of failed replicas. |
| `concurrent-replica-rebuild-per-node-limit` | `5` | Limits the number of concurrent replica rebuilds per node. |
| `offline-replica-rebuilding` | `false` | Determines if degraded replicas are rebuilt while the volume is detached. |
```

Author: Modified. Thanks.

### Settings Trade-Off Analysis

- **`fast-replica-rebuild-enabled`**
  - `enabled`  
    Skips snapshot data transfer if checksums are up to date — fast rebuild, but data isn't re-validated.
  - `disabled`  
    Performs delta rebuilding using block comparisons — slower, but ensures snapshot data integrity.

> - **[snapshot-data-integrity](../data-integrity/snapshot-data-integrity-check)**

Author: Added. Thanks.

---
title: Replica Rebuilding
weight: 6
---
- [Rebuilding Types](#rebuilding-types)
  - [Full Replica Rebuilding](#full-replica-rebuilding)
  - [Delta Replica Rebuilding](#delta-replica-rebuilding)
    - [Fast Replica Rebuilding](#fast-replica-rebuilding)
- [Operations That Affect Rebuilding Performance](#operations-that-affect-rebuilding-performance)
- [Relevant Settings](#relevant-settings)

When Longhorn detects a failed replica or a replica deleted, it automatically triggers a rebuilding process. This document introduces the new replica rebuilding flow, designed to improve performance and support delta rebuilding and fast rebuilding optimizations.

Replica rebuilding may be triggered under the following scenarios:

- Node reboot or eviction.
- Replica becomes unhealthy or deleted.

## Rebuilding Types

| Type | Description |
| --- | --- |
| Full Rebuilding | The replica contains no valid data or is deleted and must sync the entire volume. |

> | Delta Rebuilding | The replica is partially reusable; only changed blocks are synchronized. |

Author: Removed. Thanks.

|||

 {{< figure alt="Replica Rebuilding Flow Diagram" src="/img/diagrams/architecture/replica-rebuilding-flow.png" >}}

### Full Replica Rebuilding

If the replica is unrecoverable or replenished:

- Enter full rebuild mode.

> - Reconstruct the replica by sending the full snapshot chain.

Author: Modified. Thanks

### Delta Replica Rebuilding

1. Mark the target replica with `WO` (Write-Only) mode.

2. System will create a new snapshot as a reference point of the volume-head for integrity check.

3. Get Synchronizing File List of the volume head and snapshot files for checksum comparison.

4. Compare Checksums. If [Fast Replica Rebuilding](#fast-replica-rebuilding) is not enabled, use interval-based checksum comparison.

5. If local and remote checksums do not match, launch receiver of the target replica to synchronize missing or mismatched data.

> #### Fast Replica Rebuilding

Author: Modified. Thanks

Fast rebuilding is enabled when:

- `fast-replica-rebuild-enabled: true`
- Snapshot checksums are created.

  - `snapshot-data-integrity`: `enabled` or
  - `snapshot-data-integrity-immediate-check-after-snapshot-creation: true`

Benefits:

- Skip checking unchanged snapshots.
- Reduce network IO and rebuilding time.

For more details, [Fast Replica Rebuilding](./fast-replica-rebuilding).

> ## Operations That Affect Rebuilding Performance

Author: Modified. Thanks
